content/blog/2019-10-08-remote-read-meets-streaming.md

```diff
@@ -0,0 +1,313 @@
+* Prometheus being able to read from 3rd party long term storage systems e.g InfluxDB.
+* 3rd party system querying data from Prometheus e.g [Thanos](https://thanos.io).
+
+The remote read API exposes a simple HTTP endpoint that expects following protobuf payload:
+
+```
+message Query {
+  int64 start_timestamp_ms = 1;
+  int64 end_timestamp_ms = 2;
+  repeated prometheus.LabelMatcher matchers = 3;
+  prometheus.ReadHints hints = 4;
+}
+```
+
+With this payload, the client can request certain series matching given `matchers` and time range with `end` and `start`.
+
+The response is equally simple:
+
+```
+message Sample {
+  double value    = 1;
+  int64 timestamp = 2;
+}
+
+message TimeSeries {
+  repeated Label labels   = 1;
+  repeated Sample samples = 2;
+}
+
+message QueryResult {
+  repeated prometheus.TimeSeries timeseries = 1;
+}
+```
+
+Remote read returns the matched time series with **raw** samples of value and timestamp.
+
+## Problem Statement
+
+There were two key problems for such a simple remote read. It was easy to use and understand, but there were no
+streaming capabilities within single HTTP request for the protobuf format we defined. Secondly, the response was 
+including raw samples (`float64` value and `int64` timestamp) instead of
+an encoded, compressed batch of samples called "chunks" that are used to store metrics inside TSDB.
+
+The server algorithm for remote read before improvement was: 
+
+1. Parse request.
+1. Select metrics from TSDB.
+1. For all decoded series:
+  * For all samples:
+      * Add to response protobuf
+1. Marshal response.
+1. Snappy compress.
+1. Send back the HTTP response.  
+ 
+The whole response of the remote read had to be buffered in raw, uncompressed form to marshal it 
+in a potentially huge protobuf before sending it to the client. The whole response has to then be fully buffered in the client again to be able
+to unmarshal it from protobuf. Only after that client could access raw samples.
+
+What does it mean? It means that requests for, let's say, only 8 hours that matches 10,000 series can take up to **2.5GB** of memory allocated by both client and server each!
+
+Find below, memory usage metric for both Prometheus and [Thanos Sidecar](https://thanos.io/components/sidecar.md/) (remote read client) during remote read request time:
+
+![Prometheus 2.12.0: RSS of single read 8h of 10k series](/assets/blog/2019-10-08/10kseries8hours-2.12.png)
+
+![Prometheus 2.12.0: Heap-only allocations of single read 8h of 10k series](/assets/blog/2019-10-08/10series8hours-2.12-allocs.png)
+
+It is worth to noting that querying 10,000 series is not a great idea, even for Prometheus native HTTP `query_range` endpoint,
+as your browser simply will not be happy fetching, storing and rendering hundreds of megabytes of data. Additionally,
+for dashboards and rendering purposes it is not practical to have that much data. That's why usually we craft queries that have no more than 20 series.
```

Comment:
This implies a performance issue with Prometheus, I'd reframe this as an issue for humans reading the graphs.

Reply:
:+1: